README.md
# Return Target Icon Bar

![UI](https://i.gyazo.com/6ad69acaa913d1aa08a4b767d6ac0b0c.png)

## How to use:

### Left mouse on raid symbol:
Search for a mob with the specified symbol. The addon searches all raid or party members and their current targets for the specified symbol. If a unit is found that contains the specific symbol it will be targeted.

If no unit is found an error sound is played.

### Shift + Left mouse on raid symbol:
Set a raid symbol to your actual target, if the actual target already has the specified symbol, the symbol will be removed.


## Configuration:

![Configuration](https://i.gyazo.com/76e03d8c661fd3944f98a7f44906031f.png)

To open the options dialog use /ReturnTargetIconBar or /rtib.

### Key Bindings can be set via "Key Bindings -> Addons"
![Configuration](https://i.gyazo.com/17983d7649a92e26e168e213d127edde.png)


##

This addon is based on the vanilla addon BananaBar: https://github.com/webdkpthroaway/banana

CHANGES.txt
tag 3f734ff7555c87ee1f7c5fff973882c7c5388931 0.3.0
Author:	zorkqz <zorkqz@gmail.com>
Date:	Thu Dec 26 20:12:50 2019 +0100

commit 52efe328551c9540e37ec98858c59c0619165a06
Author: zorkqz <zorkqz@gmail.com>
Date:   Thu Dec 26 20:10:24 2019 +0100

    Add Key Bindings

commit 52efe328551c9540e37ec98858c59c0619165a06
Author: zorkqz <zorkqz@gmail.com>
Date:   Thu Dec 26 20:10:24 2019 +0100

    Add Key Bindings

commit fe0f56623d360e712bcc1f555d90054f0c3abbeb
Author: zorkqz <zorkqz@gmail.com>
Date:   Thu Dec 26 19:16:34 2019 +0100

    Update config image in readme

commit ca1d90c301767cef2cd5d069e357774614bcd10e
Author: zorkqz <zorkqz@gmail.com>
Date:   Thu Dec 26 19:14:30 2019 +0100

    Allow hiding the text

commit 3199691b5545801911e97b8ed67e835ef38c5b2b
Author: zorkqz <zorkqz@gmail.com>
Date:   Thu Dec 26 19:01:27 2019 +0100

    Update Interface Version
    
    (cherry picked from commit 6d86b27b350fad51ad47eb0b1434c31052b69b76)

commit caedd414c951b1799c5dce01836a690af9a5ee5b
Author: zorkqz <zorkqz@gmail.com>
Date:   Tue Dec 10 18:40:55 2019 +0100

    Update toc version



Config.lua
local ZkyTargetBar = LibStub("AceAddon-3.0"):GetAddon("ZkyTargetBar")

local defaults = {
    profile = {
        position = nil,
        scale = 1.5,
        layout = 3,
        textPosition = { x = 0, y = -2 },
        hideText = false,
        showInRaid = true,
        showInParty = false,
        showSolo = false,
        locked = false,
        playErrorSound = true,
    }
}

local options = nil
local function getOptions()
    if options then return options end

    options = {
        type = "group",
        name = "Return Target Icon Bar",
        args = {
            general = {
                type = "group",
                name = "",
                inline = true,
                args = {
                    descriptionImage = {
                        name = "",
                        type = "description",
                        image = "Interface\\AddOns\\ZkyTargetBar\\media\\return",
                        imageWidth = 64,
                        imageHeight = 64,
                        width = "half",
                        order = 1
                    },
                    descriptiontext = {
                        name = "Return Target Icon Bar by Irpa\nDiscord: https://discord.gg/SZYAKFy\nGithub: https://github.com/zorkqz/ZkyTargetBar\n",
                        type = "description",
                        width = "full",
                        order = 1.1
                    },
                    headerGlobalSettings = {
                        name = "Settings",
                        type = "header",
                        width = "double",
                        order = 2
                    },
                    showInRaid = {
                        name = "Show in raids",
                        desc = "",
                        type = "toggle",
                        order = 2.1,
                        get = function(self)
                            return ZkyTargetBar.db.profile.showInRaid
                        end,
                        set = function(self, value)
                            ZkyTargetBar.db.profile.showInRaid = value
                            ZkyTargetBar:CheckHide()
                        end
                    },
                    showInParty = {
                        name = "Show in party",
                        desc = "",
                        type = "toggle",
                        order = 2.2,
                        get = function(self)
                            return ZkyTargetBar.db.profile.showInParty
                        end,
                        set = function(self, value)
                            ZkyTargetBar.db.profile.showInParty = value
                            ZkyTargetBar:CheckHide()
                        end
                    },
                    showSolo = {
                        name = "Show when solo",
                        desc = "Doesn't make much sense.",
                        type = "toggle",
                        order = 2.3,
                        get = function(self)
                            return ZkyTargetBar.db.profile.showSolo
                        end,
                        set = function(self, value)
                            ZkyTargetBar.db.profile.showSolo = value
                            ZkyTargetBar:CheckHide()
                        end
                    },
                    locked = {
                        name = "Lock position",
                        desc = "Lock icons in place.",
                        type = "toggle",
                        order = 2.4,
                        get = function(self)
                            return ZkyTargetBar.db.profile.locked
                        end,
                        set = function(self, value)
                            ZkyTargetBar.db.profile.locked = value
                        end
                    },
                    playErrorSound = {
                        name = "Play error sound",
                        desc = "Play an error sound if no target could be found.",
                        type = "toggle",
                        order = 2.4,
                        get = function(self)
                            return ZkyTargetBar.db.profile.playErrorSound
                        end,
                        set = function(self, value)
                            ZkyTargetBar.db.profile.playErrorSound = value
                        end
                    },
                    hideText = {
                        name = "Hide Text",
                        desc = "",
                        type = "toggle",
                        order = 2.5,
                        get = function(self)
                            return ZkyTargetBar.db.profile.hideText
                        end,
                        set = function(self, value)
                            ZkyTargetBar.db.profile.hideText = value
                            ZkyTargetBar:UpdateTextPosition()
                        end
                    },
                    headerLayout = {
                        name = "Layout",
                        type = "header",
                        width = "double",
                        order = 3
                    },
                    scale = {
                        name = "Scale",
                        desc = "",
                        type = "range",
                        order = 3.1,
                        min = 0.1,
                        max = 5,
                        softMin = 0.5,
                        softMax = 2,
                        step = 0.1,
                        bigStep = 0.1,
                        get = function(self)
                            return ZkyTargetBar.db.profile.scale
                        end,
                        set = function(self, value)
                            ZkyTargetBar.db.profile.scale = value
                            ZkyTargetBar:DoLayout()
                        end
                    },
                    layout = {
                        name = "Layout",
                        desc = "",
                        type = "range",
                        order = 3.2,
                        min = 0,
                        max = 3,
                        softMin = 0,
                        softMax = 3,
                        step = 1,
                        bigStep = 1,
                        get = function(self)
                            return ZkyTargetBar.db.profile.layout
                        end,
                        set = function(self, value)
                            ZkyTargetBar.db.profile.layout = value
                            ZkyTargetBar:DoLayout()
                        end
                    },
                    emptyDescription1 = {
                        name = "",
                        type = "description",
                        order = 3.3
                    },
                    textPositionX = {
                        name = "Text X",
                        desc = "",
                        type = "range",
                        order = 3.4,
                        min = -100,
                        max = 100,
                        softMin = -100,
                        softMax = 100,
                        step = 0.1,
                        bigStep = 1,
                        get = function(self)
                            return ZkyTargetBar.db.profile.textPosition.x
                        end,
                        set = function(self, value)
                            ZkyTargetBar.db.profile.textPosition.x = value
                            ZkyTargetBar:UpdateTextPosition()
                        end
                    },
                    textPositionY = {
                        name = "Text Y",
                        desc = "",
                        type = "range",
                        order = 3.5,
                        min = -100,
                        max = 100,
                        softMin = -100,
                        softMax = 100,
                        step = 0.1,
                        bigStep = 1,
                        get = function(self)
                            return ZkyTargetBar.db.profile.textPosition.y
                        end,
                        set = function(self, value)
                            ZkyTargetBar.db.profile.textPosition.y = value
                            ZkyTargetBar:UpdateTextPosition()
                        end
                    },
                    
                }
            }
        }
    }

    return options
end

function ZkyTargetBar:ChatCommand(input)
    InterfaceOptionsFrame_OpenToCategory(ZkyTargetBar.Options.Menus.Profiles)
    InterfaceOptionsFrame_OpenToCategory(ZkyTargetBar.Options.Menus.Profiles)
    InterfaceOptionsFrame_OpenToCategory(ZkyTargetBar.Options.Menus.ZkyTargetBar)
end


function ZkyTargetBar:SetupOptions()

    ZkyTargetBar.db = LibStub("AceDB-3.0"):New("ZkyTargetBarDB", defaults, true)

    ZkyTargetBar.Options = {}
    ZkyTargetBar.Options.Menus = {}
    ZkyTargetBar.Options.Profiles = LibStub("AceDBOptions-3.0"):GetOptionsTable(ZkyTargetBar.db)
    
    LibStub("AceConfig-3.0"):RegisterOptionsTable("ZkyTargetBar", getOptions)
	LibStub("AceConfigRegistry-3.0"):RegisterOptionsTable("ZkyTargetBar Profiles", ZkyTargetBar.Options.Profiles)
    
    ZkyTargetBar.Options.Menus.ZkyTargetBar
        = LibStub("AceConfigDialog-3.0"):AddToBlizOptions("ZkyTargetBar", "ZkyTargetBar", nil)
    ZkyTargetBar.Options.Menus.Profiles
        = LibStub("AceConfigDialog-3.0"):AddToBlizOptions("ZkyTargetBar Profiles", "Profiles", "ZkyTargetBar")
    
    ZkyTargetBar:RegisterChatCommand("ZkyTargetBar", "ChatCommand")
    ZkyTargetBar:RegisterChatCommand("ztb", "ChatCommand")
    
end


-- Binding Variables
BINDING_HEADER_RETURNTARGETBAR = "Return Target Icon Bar" 
_G["BINDING_NAME_CLICK ZkyTargetBarButton8:LeftButton"] = "Target Skull";
_G["BINDING_NAME_CLICK ZkyTargetBarButton7:LeftButton"] = "Target Cross";
_G["BINDING_NAME_CLICK ZkyTargetBarButton6:LeftButton"] = "Target Square";
_G["BINDING_NAME_CLICK ZkyTargetBarButton5:LeftButton"] = "Target Moon";
_G["BINDING_NAME_CLICK ZkyTargetBarButton4:LeftButton"] = "Target Triangle";
_G["BINDING_NAME_CLICK ZkyTargetBarButton3:LeftButton"] = "Target Diamond";
_G["BINDING_NAME_CLICK ZkyTargetBarButton2:LeftButton"] = "Target Circle";
_G["BINDING_NAME_CLICK ZkyTargetBarButton1:LeftButton"] = "Target Star";


ZkyTargetBar.lua
local ZkyTargetBar = LibStub("AceAddon-3.0"):NewAddon("ZkyTargetBar", "AceConsole-3.0", "AceEvent-3.0")

local count = {}

function ZkyTargetBar:OnInitialize()
    ZkyTargetBar:SetupOptions()
    
    ZkyTargetBar.NextTime = 0
    ZkyTargetBar.MainFrame = CreateFrame("Frame", "ZkyTargetBarUI", UIParent);
    
    ZkyTargetBar.MainFrame:ClearAllPoints()
    if ZkyTargetBar.db.profile.position ~= nil then
        ZkyTargetBar.MainFrame:SetPoint("BOTTOMLEFT", ZkyTargetBar.db.profile.position.x, ZkyTargetBar.db.profile.position.y)
    else
        ZkyTargetBar.MainFrame:SetPoint("CENTER", UIParent)
    end
    
    ZkyTargetBar.MainFrame:EnableMouse(true)
    ZkyTargetBar.MainFrame:SetMovable(true)
    ZkyTargetBar.MainFrame:SetHeight(1)
    ZkyTargetBar.MainFrame:SetWidth(1)
    
    ZkyTargetBar.IconFrames = {}
	for i = 1, 8 do
		ZkyTargetBar.IconFrames[i] = ZkyTargetBar:CreateIconFrame(i, ZkyTargetBar.MainFrame)
	end
    
    ZkyTargetBar:DoLayout()
    ZkyTargetBar:CheckHide()
    ZkyTargetBar.MainFrame:SetScript("OnUpdate", ZkyTargetBar.OnUpdate)
    ZkyTargetBar:RegisterEvent("PLAYER_TARGET_CHANGED", ZkyTargetBar.Check)
    
    ZkyTargetBar.db.RegisterCallback(self, "OnProfileChanged", "RefreshConfig")
    ZkyTargetBar.db.RegisterCallback(self, "OnProfileCopied", "RefreshConfig")
    ZkyTargetBar.db.RegisterCallback(self, "OnProfileReset", "RefreshConfig")
end

function ZkyTargetBar:RefreshConfig()
    ZkyTargetBar:DoLayout()
    ZkyTargetBar:UpdateTextPosition()
    ZkyTargetBar:CheckHide()
end

function ZkyTargetBar:OnUpdate()
    local currentTime = GetTime()
    if currentTime < ZkyTargetBar.NextTime then
        return
    end
    ZkyTargetBar.NextTime = currentTime + 0.25
    
    ZkyTargetBar:Check()
    ZkyTargetBar:CheckHide()
end

function ZkyTargetBar:Check()
    ZkyTargetBar:CheckCount()
    ZkyTargetBar:CheckPlayerTarget()
end

function ZkyTargetBar:CheckCount()
    for i = 1, 8 do
        count[i] = 0
	end
    
    if ZkyTargetBar.db.profile.showInRaid and IsInRaid() then
        for i = 1, 40 do  
            ZkyTargetBar:CheckCountUnit("raid"..i)
        end
    elseif ZkyTargetBar.db.profile.showInParty and IsInGroup() then
        ZkyTargetBar:CheckCountUnit("player")
        for i = 1, 4 do  
            ZkyTargetBar:CheckCountUnit("party"..i)
        end
    elseif ZkyTargetBar.db.profile.showSolo then
        ZkyTargetBar:CheckCountUnit("player")
    end    
    
    for i = 1, 8 do
        ZkyTargetBar.IconFrames[i].Text:SetText(count[i])
	end
end

function ZkyTargetBar:CheckCountUnit(unit)
    if not UnitExists(unit) or not UnitExists(unit.."target") then
        return
    end
    
    for i = 1, 8 do
        if GetRaidTargetIndex(unit.."target") == i then
            count[i] = count[i] + 1
        end
	end
end

function ZkyTargetBar:CheckPlayerTarget()
    local size = 24 * ZkyTargetBar.db.profile.scale
    local playerTargetIconId = nil

    if (UnitExists('playertarget')) then
        -- print('UnitExists: ' .. 'true')
        local targetId = GetRaidTargetIndex("playertarget")
        if(targetId) then
            print('GetRaidTargetIndex: ' .. GetRaidTargetIndex("playertarget"))
        end
    end

    if UnitExists("playertarget") then
        playerTargetIconId = GetRaidTargetIndex("playertarget")
        if playerTargetIconId then
            ZkyTargetBar.IconFrames[playerTargetIconId].Texture:SetWidth(size / 2)
            ZkyTargetBar.IconFrames[playerTargetIconId].Texture:SetHeight(size / 2)
            ZkyTargetBar.IconFrames[playerTargetIconId]:SetBackdrop({
                edgeFile = "Interface\\Tooltips\\UI-Tooltip-Border",
                edgeSize = 16,
                insets = {left = 4, right = 4, top = 4, bottom = 4},
            })
        end
    end
    
    for i = 1, 8 do
        if i ~= playerTargetIconId then
            ZkyTargetBar.IconFrames[i].Texture:SetWidth(size)
            ZkyTargetBar.IconFrames[i].Texture:SetHeight(size)
            ZkyTargetBar.IconFrames[i]:SetBackdrop(nil)
        end
        ZkyTargetBar.IconFrames[i].Text:SetText(count[i])
	end
end


function ZkyTargetBar:DoLayout()
    
    local x = 0
    local y = 0
    
    local nn = math.pow(2, (ZkyTargetBar.db.profile.layout or 3) % 4)
    
    local size = 24 * ZkyTargetBar.db.profile.scale or 1
                
    for i = 1, 8 do
		local iconFrame = ZkyTargetBar.IconFrames[i];
        iconFrame:SetWidth(size)
        iconFrame:SetHeight(size)
        iconFrame.Texture:SetWidth(size)
        iconFrame.Texture:SetHeight(size)
        
        iconFrame:SetPoint("CENTER", ZkyTargetBar.MainFrame, x, y)
        x = x - size
        if i % nn == 0 then
            x = 0
            y = y + size
        end
	end
end

function ZkyTargetBar:Target(iconId)

    for i = 1, 40 do
        if ZkyTargetBar:TargetByUnit("raid"..i, iconId) then
            return;
        end
    end
    
    for i = 1, 4 do
        if ZkyTargetBar:TargetByUnit("party"..i, iconId) then
            return;
        end
    end

    if ZkyTargetBar:TargetByUnit("player", iconId) then
        return;
    end
  
    if ZkyTargetBar.db.profile.playErrorSound then
        PlaySound(846)
    end
end

function ZkyTargetBar:TargetByUnit(unit, iconId)
        
    if UnitExists(unit) and GetRaidTargetIndex(unit) == iconId then
		ZkyTargetBar.IconFrames[iconId]:SetAttribute("macrotext", "/target " .. unit)
        return 1;
    end
            
    if UnitExists(unit.."target") and GetRaidTargetIndex(unit.."target") == iconId then
		ZkyTargetBar.IconFrames[iconId]:SetAttribute("macrotext", "/target " .. unit.."target")
        return 1;
    end
    
    return nil;
end


function ZkyTargetBar:CreateIconFrame(iconId)

    local iconFrame = CreateFrame("Button", "ZkyTargetBarButton" .. iconId, ZkyTargetBar.MainFrame, "SecureUnitButtonTemplate")

    iconFrame:SetPoint("CENTER", ZkyTargetBar.MainFrame)
    iconFrame:SetWidth(24)
    iconFrame:SetHeight(24)

    iconFrame.Texture = iconFrame:CreateTexture(nil, "OVERLAY")
    iconFrame.Texture:SetWidth(24)
    iconFrame.Texture:SetHeight(24)
    iconFrame.Texture:SetPoint("CENTER", iconFrame, "CENTER", 0, 0)
    iconFrame.Texture:SetTexture("Interface\\TargetingFrame\\UI-RaidTargetingIcons")   
    SetRaidTargetIconTexture(iconFrame.Texture, iconId)
        
    iconFrame.Text = iconFrame:CreateFontString(nil, "OVERLAY", "GameFontHighlightSmall")
    iconFrame.Text:SetPoint("TOP", iconFrame, "BOTTOM", ZkyTargetBar.db.profile.textPosition.x, ZkyTargetBar.db.profile.textPosition.y)
    if ZkyTargetBar.db.profile.hideText then
        iconFrame.Text:Hide();
    end
    
	iconFrame:SetMovable(true)
	iconFrame:SetUserPlaced(true)
	iconFrame:SetAttribute("type", "macro")
    iconFrame:RegisterForClicks("LeftButtonDown")
                
    iconFrame:SetScript("OnMouseDown", function(self)
        if not ZkyTargetBar.db.profile.locked then
            ZkyTargetBar.MainFrame:StartMoving()
        end
        if IsShiftKeyDown() then
            SetRaidTarget("target", iconId)
        else
            ZkyTargetBar:Target(iconId)
        end
    end)
    iconFrame:SetScript("OnMouseUp", function(self)
        ZkyTargetBar.MainFrame:StopMovingOrSizing()
        
        if not ZkyTargetBar.db.profile.position then
            ZkyTargetBar.db.profile.position = {}
        end
        ZkyTargetBar.db.profile.position.x = ZkyTargetBar.MainFrame:GetLeft()
        ZkyTargetBar.db.profile.position.y = ZkyTargetBar.MainFrame:GetBottom()
    end)

    return iconFrame;
end

function ZkyTargetBar:UpdateTextPosition()
    for i = 1, 8 do
		local iconFrame = ZkyTargetBar.IconFrames[i];
        if ZkyTargetBar.db.profile.hideText then
            iconFrame.Text:Hide();
        else
            iconFrame.Text:Show();
            iconFrame.Text:SetPoint("TOP", iconFrame, "BOTTOM", ZkyTargetBar.db.profile.textPosition.x, ZkyTargetBar.db.profile.textPosition.y)
        end
    end
end

function ZkyTargetBar:CheckHide()

    local show = false

    if IsInRaid() then
        show = ZkyTargetBar.db.profile.showInRaid
    elseif IsInGroup() then
        show = ZkyTargetBar.db.profile.showInParty
    else
        show = ZkyTargetBar.db.profile.showSolo
    end

    if show then
        ZkyTargetBar.MainFrame:Show();
    else
        ZkyTargetBar.MainFrame:Hide();
    end
end
